Add reference Web GPU Spinning Cube Demo and glMatrix

JusSn opened the pull request:
This demo serves as a reference for a Web GPU implementation. It reflects the current API as well as approved pull requests.

Also included is the [glMatrix](http://glmatrix.net) matrix math library by [Brandon Jones](https://github.com/toji).

kainino0x commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,309 @@
+<body>
+<canvas height=1000 width=1000></canvas>
+<script>
+const shader = `
```
As before, before merge, need a big scary warning explaining the state of shader things: This is a preview of WHLSL, but the group hasn't decided what shading language will be taken in by the API.

JusSn replied:
Updated!

kainino0x replied:
Thanks! LGTM

kainino0x replied:
(on this fix. sorry, haven't reviewed the rest yet)

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+
+    const verticesBufferDescriptor = {
+        size: vertexArray.byteLength,
+        usage: GPUBufferUsage.VERTEX | GPUBufferUsage.TRANSFER_DST
```
It's kind of unfortunate that setSubData requires TRANSFER_DST and nothing else. On UMA, setSubData might be able to write directly into the buffer (if the stars align correctly).

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+<script>
+/*******
+* This preview of Web GPU uses WHLSL as its shading language. 
+* The choice of shading language(s) that will be ingested by the API is still under deliberation. 
```
This is true, but isn't relevant for this particular patch. We're attempting to make a "Hello World" type demo using best practices for WebGPU. The best practices are to use WHLSL. Tests in the test suite will use WHLSL and won't have a disclaimer, so there's no reason to have one here. We should remove the disclaimer.

grorg replied:
I think the disclaimer is less about the shading language and more about the API for ingesting the shader.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+        depthCompare: "less"
+    };
+
+    /* Render Pipeline */
```
I'd move this comment to above line 150.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+        clearColor: { r: 0.5, g: 1.0, b: 1.0, a: 1.0 }
+    };
+
+    // GPUExtent3D
```
I wish we could do something more type-safe here. It seems silly to attach a 3D size to a 2D texture.

grorg replied:
Yeah, maybe we need GPUSize2D and GPUSize3D (I prefer both of them over "Extent")

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+        mipLevelCount: 1,
+        sampleCount: 1,
+        dimension: "2d",
+        format: "d32-float-s8-uint",
```
At a glance, this looks like a UUID. 🤮

grorg replied:
We should at least make them consistent with the other texture formats.

JusSn replied:
I think I missed this one when updating the names.

JusSn replied:
Oh, you forgot to include depth and stencil formats when you updated the list of texture formats in the IDL. I'll make a PR in a bit to add them.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+        clearDepth: 1.0
+    };
+
+    renderPassDescriptor = {
```
Instead of writing into globals, it's probably better style to return a collection of all the variables.

JusSn replied:
That sounds very functional indeed.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+    };
+}
+
+class MappedBufferBindGroup {
```
Cool

grorg replied:
I think if we're going to add classes as helpers, we could do the same for the transform array matrix. e.g. a class that holds the mappedGroup and exposes the arrayBuffer.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+    queue.submit([commandEncoder.finish()]);
+
+    // Ready the current buffer for update after GPU is done with it.
+    mappedGroup.buffer.mapWriteAsync().then((arrayBuffer) => {
```
The right thing will happen here, but we might not wants apps to rely on the specific timeline that the mapWriteAsync promise resolves. It may be better to explicitly handle the dependency of "the mapping should happen after the command buffer finishes" by using fences. I'm not sure which behavior we want to promote.

grorg replied:
If this does the right thing, it's probably more readable than barriers. 

The thing that I think is a bit weird is that mappedGroups could grow infinitely (if some other part of the code appends to it outside this loop). It's a shame we can't attach the group to the queue or something.

JusSn replied:
I have this example working with my mapWriteAsync() prototype. I could use a class to control access to the mappedGroups queue, if that's what you mean?

JusSn replied:
Oh, I understand what you're saying now. Shouldn't mapWriteAsync only resolve after the GPU is done with the buffer?

magcius replied:
Yes, but it means that we don't have any pipelining here based on how I understand it. We cannot upload the data for the second frame until the GPU is done with the data for the first. Our example should support double-buffering.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+    mappedGroup.buffer.unmap();
+
+    const commandEncoder = device.createCommandEncoder();
+    renderPassDescriptor.colorAttachments[0].attachment = swapChain.getTexture().createDefaultTextureView();
```
You've got a single render pass descriptor that every render pass uses. This implicitly requires that beginRenderPass copy the contents of the descriptor so that the next frame doesn't clobber the current frame's execution. This might be okay, but it might not be. If it isn't okay, the group should decide on a mechanism to make it obvious to authors what is going wrong when they write the same code you have here.

litherum commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+
+const vertexSize = 4 * 8;
+const colorOffset = 4 * 4;
+const vertexArray = new Float32Array([
```
I have no idea whether this giant array is correct. Can't you compute it at runtime?

grorg replied:
This is fairly standard in all spinning cube demos. I don't think it is a big deal. Computing it at runtime would probably be an equally confusing set of for loops.

Maybe the best thing to do would be to name the variables cubeVertices, and provide comments for the 4 * 8 and 4 * 4 values.

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+<script src="gl-matrix-min.js"></script>
+</head>
+<body>
+<canvas height=1000 width=1000></canvas>
```
You set a 600px viewport above, so 1000x1000 would be mostly offscreen.

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+
+vertex FragmentData vertex_main(float4 position : attribute(0),
+                                float4 color : attribute(1),
+                                constant float4x4* modelViewProjectionMatrix : register(b0)) {
```
Why is the input a pointer to a float4x4?

JusSn replied:
Do you mean as opposed to a reference or copy?

magcius replied:
As opposed to the more normal HLSL mechanism of defining a cbuffer/ConstantBuffer with your input variables.

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+
+    const depthTexture = device.createTexture(depthTextureDescriptor);
+
+    // GPURenderPassDepthStencilAttachmentDescriptor
```
I don't understand this comment.

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+        this.bindGroup = bindGroup;
+    }
+}
+const mappedGroups = [];
```
This should probably go at the top.

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+        const { buffer, arrayBuffer } = device.createMappedBuffer(transformBufferDescriptor);
+        const bindGroup = device.createBindGroup(createTransformBindGroupDescriptor(buffer));
+
+        mappedGroup = new MappedBufferBindGroup(buffer, arrayBuffer, bindGroup);
```
Since a MappedBufferBindGroup doesn't have anything other than a constructor, it's basically equivalent to an anonymous object. i.e. you could omit the class and just say:

mappedGroup = { buffer, arrayBuffer, bindGroup };

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+    // Bind groups
+    passEncoder.setBindGroup(0, mappedGroup.bindGroup);
+
+    passEncoder.draw(36, 1, 0, 0);
```
You should explain where the 36 came from. And probably the 1, 0, 0 part too.

JusSn replied:
Done.

grorg commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+init();
+</script>
+</body>
+</html>
```
Add a newline.

JusSn commented on samples/hello-cube.html:
```diff
@@ -0,0 +1,313 @@
+    } else
+        mappedGroup = mappedGroups.shift();
+
+    updateTransformArray(mappedGroup.arrayBuffer);
```
Leaving myself a note here to fix: updateTransformArray takes a Float32Array, not an ArrayBuffer.

JusSn commented:
I've updated the file to match our API changes and removed the library file in favor of an external source. It does not work in current Safari since WHLSL is not the default ingestion format. 

Feel free to make a PR after merging to feature-detect and use SPIR-V shaders, or other API updates I may have missed @kainino0x @austinEng
